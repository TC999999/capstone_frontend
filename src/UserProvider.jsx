import UserContext from "./UserContext.js";
import { useState, useEffect } from "react";
import marketAPI from "../api.js";

const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  //sets context for user for entire app
  const currentUser = (user) => {
    setUser(user);
  };

  // //On initial render, get user token from local storage, gets the current user from the api, and set as context
  useEffect(() => {
    async function getToken() {
      let token = localStorage.getItem("market-token");
      if (token) {
        let userInfo = await marketAPI.getCurrentUser();
        currentUser(userInfo);
      }
    }
    getToken();
  }, []);

  //These functions and user state will be used in context
  //Set the token in local storage, retrives the current user from the api, and sets the user as context on login
  const logIn = async (username, token) => {
    localStorage.setItem("market-token", token);
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  //remove token and user from local storage and set user context to null on logout
  const logOut = () => {
    localStorage.removeItem("market-token");
    currentUser(null);
  };

  //update current user in local storage and context
  const updateUser = async (username) => {
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  return (
    <UserContext.Provider
      value={{ user, currentUser, logIn, logOut, updateUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
