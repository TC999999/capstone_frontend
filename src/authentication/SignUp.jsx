import { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import marketAPI from "../../api";
import UserContext from "../UserContext";
import "../styles/Signup.css";

const Signup = ({ isLoading, setIsLoading }) => {
  const initialState = {
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
    address: "",
    zipCode: "",
    city: "",
    regionOrState: "",
    country: "United States of America",
  };
  const { logIn } = useContext(UserContext);
  const navigate = useNavigate();
  const [err, setErr] = useState(false);
  const [message, setMessage] = useState("");
  const [formData, setFormData] = useState(initialState);

  //navigator.geolocation.getGurrentPosition options
  const options = {
    enableHighAccuracy: true,
    timeout: 5000,
    maximumAge: 0,
  };

  //gets the user's current location in longitude and latitude
  function getLocation() {
    return new Promise((res, rej) => {
      navigator.geolocation.getCurrentPosition(res, rej, options);
    });
  }

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({ ...data, [name]: value }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setIsLoading(true);
    let position = await getLocation();
    const { latitude, longitude } = position.coords;
    const {
      username,
      password,
      firstName,
      lastName,
      email,
      address,
      zipCode,
      city,
      regionOrState,
      country,
    } = formData;
    try {
      const token = await marketAPI.signUp({
        username,
        password,
        firstName,
        lastName,
        email,
        address,
        zipCode,
        city,
        regionOrState,
        country,
        latitude,
        longitude,
      });

      logIn(username, token);
      setFormData({ ...initialState });
      navigate("/");
    } catch (err) {
      setErr(true);
      setMessage(err);
      setIsLoading(false);
    }
  };

  return (
    <div className="signup-form">
      <h1>Sign Up</h1>
      <form onSubmit={handleSubmit}>
        <div className="username-div">
          <label htmlFor="username">Username: </label>
          <input
            id="signup_username"
            type="text"
            name="username"
            placeholder="type your username here"
            value={formData.username}
            onChange={handleChange}
            required
          />
        </div>
        <div className="password-div">
          <label htmlFor="password">Password: </label>
          <input
            id="signup_password"
            type="password"
            name="password"
            placeholder="type your password here"
            value={formData.password}
            onChange={handleChange}
            required
          />
        </div>
        <div className="firstName-div">
          <label htmlFor="firstName">First Name: </label>
          <input
            id="firstName"
            type="text"
            name="firstName"
            placeholder="type your first name here"
            value={formData.firstName}
            onChange={handleChange}
            required
          />
        </div>
        <div className="lastName-div">
          <label htmlFor="lastName">Last Name: </label>
          <input
            id="lastName"
            type="text"
            name="lastName"
            placeholder="type your last name here"
            value={formData.lastName}
            onChange={handleChange}
            required
          />
        </div>
        <div className="email-div">
          <label htmlFor="email">Email: </label>
          <input
            id="email"
            type="email"
            name="email"
            placeholder="type your email here"
            value={formData.email}
            onChange={handleChange}
            required
          />
        </div>
        <div className="address-div">
          <label htmlFor="address">Address: </label>
          <input
            id="address"
            type="text"
            name="address"
            placeholder="type your address here"
            value={formData.address}
            onChange={handleChange}
            required
          />
        </div>
        <div className="zipCode-div">
          <label htmlFor="zipCode">Zipcode: </label>
          <input
            id="zipCode"
            type="number"
            name="zipCode"
            placeholder="type your zipCode here"
            value={formData.zipCode}
            onChange={handleChange}
            required
          />
        </div>
        <div className="city-div">
          <label htmlFor="city">City: </label>
          <input
            id="city"
            type="text"
            name="city"
            placeholder="type your city here"
            value={formData.city}
            onChange={handleChange}
            required
          />
        </div>
        <div className="regionOrState-div">
          <label htmlFor="regionOrState">Region Or State: </label>
          <input
            id="regionOrState"
            type="text"
            name="regionOrState"
            placeholder="type your region or state here"
            value={formData.regionOrState}
            onChange={handleChange}
          />
        </div>
        <div className="country-div">
          <label htmlFor="country">Country: </label>
          <input
            id="country"
            type="text"
            name="country"
            placeholder="type your country here"
            value={formData.country}
            onChange={handleChange}
            required
          />
        </div>
        {!isLoading ? (
          <div className="button-div">
            <button className="get-profile-button">Sign Up!</button>
          </div>
        ) : (
          <div>
            <p>Loading...</p>
          </div>
        )}
      </form>
      {err && <p className="err-msg">{message}</p>}
    </div>
  );
};

export default Signup;
