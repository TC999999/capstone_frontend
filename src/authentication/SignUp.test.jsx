import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../UserProvider.jsx";
import SignUp from "./SignUp.jsx";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <SignUp />
      </UserProvider>
    </MemoryRouter>
  );
});

it("should match snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <SignUp />
      </UserProvider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

it("should have more than just username an password inputs", function () {
  const { container } = render(
    <MemoryRouter>
      <UserProvider>
        <SignUp />
      </UserProvider>
    </MemoryRouter>
  );

  const username_input = container.querySelector("#signup_username");
  const password_input = container.querySelector("#signup_password");
  const firstname_input = container.querySelector("#firstName");
  const lastname_input = container.querySelector("#lastName");

  expect(username_input).toBeInTheDocument();
  expect(password_input).toBeInTheDocument();
  expect(firstname_input).toBeInTheDocument();
  expect(lastname_input).toBeInTheDocument();
});
