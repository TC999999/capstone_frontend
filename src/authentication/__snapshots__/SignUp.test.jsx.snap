// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="signup-form"
  >
    <h1>
      Sign Up
    </h1>
    <form>
      <div
        class="username-div"
      >
        <label
          for="username"
        >
          Username: 
        </label>
        <input
          id="signup_username"
          name="username"
          placeholder="type your username here"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="password-div"
      >
        <label
          for="password"
        >
          Password: 
        </label>
        <input
          id="signup_password"
          name="password"
          placeholder="type your password here"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="firstName-div"
      >
        <label
          for="firstName"
        >
          First Name: 
        </label>
        <input
          id="firstName"
          name="firstName"
          placeholder="type your first name here"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="lastName-div"
      >
        <label
          for="lastName"
        >
          Last Name: 
        </label>
        <input
          id="lastName"
          name="lastName"
          placeholder="type your last name here"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="email-div"
      >
        <label
          for="email"
        >
          Email: 
        </label>
        <input
          id="email"
          name="email"
          placeholder="type your email here"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="address-div"
      >
        <label
          for="address"
        >
          Address: 
        </label>
        <input
          id="address"
          name="address"
          placeholder="type your address here"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="zipCode-div"
      >
        <label
          for="zipCode"
        >
          Zipcode: 
        </label>
        <input
          id="zipCode"
          name="zipCode"
          placeholder="type your zipCode here"
          required=""
          type="number"
          value=""
        />
      </div>
      <div
        class="city-div"
      >
        <label
          for="city"
        >
          City: 
        </label>
        <input
          id="city"
          name="city"
          placeholder="type your city here"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="regionOrState-div"
      >
        <label
          for="regionOrState"
        >
          Region Or State: 
        </label>
        <input
          id="regionOrState"
          name="regionOrState"
          placeholder="type your region or state here"
          type="text"
          value=""
        />
      </div>
      <div
        class="country-div"
      >
        <label
          for="country"
        >
          Country: 
        </label>
        <input
          id="country"
          name="country"
          placeholder="type your country here"
          required=""
          type="text"
          value="United States of America"
        />
      </div>
      <div
        class="button-div"
      >
        <button
          class="get-profile-button"
        >
          Sign Up!
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
